### ISSUE TITLE
[consistent use] `F` vs. `nn.functional`

### ISSUE THREAD

**Issue by stas00** [huggingface/transformers#11600]
We use 3 different ways of doing the same:
1. `F.foo()`
2. `nn.functional.foo()`
3. `torch.nn.functional.foo()`

and these could also be imported:

4. `from torch.nn.functional import foo; foo()`

Asking others it appears that `F` is not quite liked, so it's 2, 3 or 4.

2 and 3 often lead to longer lines which autoformatter wraps, leading to 3 lines of code instead of 1 and which gives less readable code.

So it seems that option 4 might be the best outcome.

For 2, the global update would be easy:
```
find . -type d -name ".git" -prune -o -type f -exec perl -pi -e 's|from torch.nn import functional as F||'  {} \;
find . -type d -name ".git" -prune -o -type f -exec perl -pi -e 's|import torch.nn.functional as F||'  {} \;
find . -type d -name ".git" -prune -o -type f -exec perl -pi -e 's| F\.| nn.functional.|g'  {} \;
make fixup
```

For 4, it will take much more work, but can be semi-automated.

@LysandreJik, @sgugger, @patrickvonplaten

**Comment by sgugger** [comment]
I personally vote for 4, and if there is any risk of ambiguity, doing
```
from torch.nn.functional import foo as torch_foo
```

**Comment by patrickvonplaten** [comment]
2 or 4 for me, but I don't have strong feeling about this issue

**Comment by LysandreJik** [comment]
2 looks good to me but no strong feelings either

**Comment by stas00** [comment]
I further analyzed the codebase and it looks like we use `nn.` almost exclusively, for all other cases outside of `nn.functional`, with an occasional `torch.nn.` 

So I'd say 2 is probably the most consistent way, unless we start explicitly importing `Parameter`, `Embedding`, `ModuleList` from `nn`.

I know the additional "functional" string pushes some single lines into autoformatter's 3 lines mode. But for consistency I think sticking to just a single:
```
from torch import nn
```
is a goodness.

Sylvain, will option 2 work for you?

**Comment by stas00** [comment]
If we stick to 2, the I will also normalize `s|torch.nn.foo|nn.foo|` calls unless you object.

**Comment by sgugger** [comment]
That works for me.

**Comment by stas00** [comment]
Status:

merged:
- [x] src

todo:
- [ ] examples
- [ ] templates
- [ ] tests

**Comment by stas00** [comment]
just need to merge docs, but all is done here.
